### PR TITLE
Added symlink support

### PR DESCRIPTION
**Pull request by gerow**
There are probably still a few quirks with the way symlinks are being handled, but it seems mostly solid from my testing.

### CODE REVIEW

**Comment by gerow** [comment]
I went ahead and cleaned up some of the debug messages. PTAL.

File: kernel/fs/vfs.c [hunk 1]
```diff
@@ -804,13 +840,80 @@ fs_node_t *kopen(char *filename, uint32_t flags) {
...
+			char symlink_buf[MAX_SYMLINK_SIZE];
+			int len = node_ptr->readlink(node_ptr, symlink_buf, sizeof(symlink_buf));
+			if (len < 0) {
+				/* TODO(gerow): should probably be setting errno from this */
+				debug_print(WARNING, "Got error %d from symlink for %s.", len, node_ptr->name);
+				free((void *)path);
+				close_fs(node_ptr);
+				return NULL;
+			}
+			if (symlink_buf[len - 1] != '\0') {
```
Comment (klange) [inline kernel/fs/vfs.c]: `len` doesn't include the '\0' (we return `3` for a symlink that points to `vim`).

File: kernel/fs/vfs.c [hunk 2]
```diff
@@ -804,13 +840,80 @@ fs_node_t *kopen(char *filename, uint32_t flags) {
...
+				/* TODO(gerow): should probably be setting errno from this */
+				debug_print(WARNING, "readlink for %s doesn't end in a null pointer. That's weird...", len, node_ptr->name);
```
Comment (klange) [inline kernel/fs/vfs.c]: Really need to get `printf` warnings set up for the kernel... You've got a `len` as your first variable argument, but only a `%s` in your format string. This was disguising the failure above when I was trying to debug it, as I was getting an in-kernel segfault trying to dereference `0x00000003` ^_^

File: kernel/include/ext2.h [hunk 3]
```diff
@@ -143,7 +143,10 @@ struct ext2_inodetable {
 	uint32_t blocks;
 	uint32_t flags;
 	uint32_t osd1;
-	uint32_t block[15];
+	union {
```
Comment (klange) [inline kernel/include/ext2.h]: No anonymous unions in C99. I took the unfortunately ugly approach of leaving this as just `block` and using a macro to get the pseudo-`symlink` member, which meant I had to do even more terrible things where you were taking `sizeof(inode->symlink)`, and I am not proud of this :(

File: kernel/include/ext2.h [hunk 4]
```diff
@@ -152,6 +155,7 @@ struct ext2_inodetable {
 } __attribute__ ((packed));
 
 typedef struct ext2_inodetable ext2_inodetable_t;
+_Static_assert(sizeof(ext2_inodetable_t) == 128, "assert ext2_inodetable_t is 128 bytes");
```
Comment (klange) [inline kernel/include/ext2.h]: No static assert in C99 either.

File: userspace/core/ls.c [hunk 5]
```diff
@@ -84,6 +85,9 @@ static void print_entry(struct tfile * file, int colwidth) {
 	if (S_ISDIR(file->statbuf.st_mode)) {
 		/* Directory */
 		ansi_color_str = DIR_COLOR;
+	} else if (S_ISLNK(file->statbuf.st_mode)) {
```
Comment (klange) [inline userspace/core/ls.c]: There's some dumb duplication of the coloring between short and long output modes in `ls` and you missed the other one. Should probably deduplicate it at some point...

**Comment by klange** [comment]
Bunch of inline comments on things I've "corrected" (some of my fixes as I'm merging them are terrible and should be addressed), but we're merging it!

File: kernel/include/ext2.h (hunk 3 above)
Comment (gerow) [inline kernel/include/ext2.h]: There's also the option of using a named union named something like sb (symlink/block). Makes getting at the block require inode->sb.block, but eh.

Comment (klange) [inline kernel/include/ext2.h]: Yeah, of the two, that's probably the better approach, but there were fewer references to inode->symlink...

File: kernel/fs/vfs.c (hunk 1 above)
Comment (klange) [inline kernel/fs/vfs.c]: This probably implies that our symlink creation is setting the size to one extra byte...

Comment (gerow) [inline kernel/fs/vfs.c]: Yup. From what I gathered readlink is supposed to null-terminate the buffer, so I just assumed ext2 null terminated the symlink on-disk. That's probably not actually the case...

**Pull request closed without merging**